### ISSUE TITLE
Advice on debugging a failure.

### ISSUE THREAD

**Issue by ckaran** [camshaft/bolero#18]
I need some advice on how to track down a failure that I strongly suspect is coming from bolero.  I have 3 tests that consistently fail, but they don't produce any failure output; they just get marked as failures.  My suspicion is that its because they are running for more than 60 seconds, and that deep in the bowels of bolero they aren't producing enough output, so that's why the test is failing.  However, I don't have a good way of verifying this, and there is too much code for me to produce a reduced example.  Do you have any suggestions?

Here is an example of the cleaned up output:

```bash
cfkaran2@Hammer:~/Documents/repositories/bit_network$ RUST_BACKTRACE=full cargo test -- configuration_parsing::scenario_parser::tests::_config_tests::try_from_scenario

<Removed all the compilation results>

running 1 test
test configuration_parsing::scenario_parser::tests::_config_tests::try_from_scenario ... FAILED

failures:

failures:
    configuration_parsing::scenario_parser::tests::_config_tests::try_from_scenario

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 227 filtered out

error: test failed, to rerun pass '--lib'
cfkaran2@Hammer:~/Documents/repositories/bit_network$ 
```

This happens in both debug and release mode.  I have all debug symbols on.  There actually are two lines with `failures` in it, that isn't a copy/paste error.

**Comment by ckaran** [comment]
Well, I've modified my code so that all collection objects have something like `#[generator(HashSet::gen().with().len(0usize..5))]` set on them.  This ensures that no test runs for over 60 seconds, but that still didn't fix the problem.  And I still don't get any output from the failed tests, so I have no idea what's going on that's causing the failure.    Any advice would be appreciated!

**Comment by ckaran** [comment]
OK, I have a suspicion, but need confirmation.  I remember that bolero has to search for symbols, which is why you need all the debug symbols turned on; how does it handle two symbols that have the same name, but are in different paths?  I have several structs defined in the same crate that have the same name (`Message`), but which are on different paths.  Rust is OK with this because the namespaces are different, but I don't know if the fuzzers/bolero can correctly handle this situation.  Note that I do import them into the same file in several places (using `as` to create new names for them).  I have no idea if that is the problem, it just seemed plausible to me.

**Comment by ckaran** [comment]
Interesting new find; running `cargo test` and `cargo bolero fuzz` yield different output sometimes.  I don't mean that bolero finds new results, I mean quite literally, the former fails with no output, while the latter will always yield some amount of output.  Not sure what is going on there, and although I've now fully refactored my code in preparation for creating a reduced test crate, I haven't had a chance to do so yet.  Once I'm sure I can reproduce this bug and produce a small test crate, I'll either update this issue, or close it out in favor of a better named one.

**Comment by ckaran** [comment]
I finally had a failure that might some shed light on this:

```bash
cargo bolero fuzz -t 100sec communications::messages::message::tests::merge
    Finished test [unoptimized + debuginfo] target(s) in 0.07s

running 1 test
WARNING: Failed to find function "__sanitizer_acquire_crash_state".
WARNING: Failed to find function "__sanitizer_print_stack_trace".
WARNING: Failed to find function "__sanitizer_set_death_callback".
INFO: Seed: 1089555166
INFO: Loaded 1 modules   (1159931 inline 8-bit counters): 1159931 [0x5600931efe00, 0x56009330b0fb), 
INFO: Loaded 1 PC tables (1159931 PCs): 1159931 [0x56009330b100,0x5600944be0b0), 
INFO:     1663 files found in /home/cfkaran2/Documents/repositories/bit_network/bit_network_library/src/communications/messages/__fuzz__/communications__messages__message__tests__merge/corpus
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: seed corpus: files: 1663 min: 1b max: 1617b total: 672085b rss: 84Mb
#1664	INITED cov: 7151 ft: 29966 corp: 535/164Kb exec/s: 832 rss: 95Mb
#1916	NEW    cov: 7151 ft: 29971 corp: 536/166Kb lim: 1577 exec/s: 638 rss: 95Mb L: 1511/1567 MS: 2 ChangeASCIIInt-ChangeBit-
#2008	REDUCE cov: 7151 ft: 29971 corp: 536/166Kb lim: 1577 exec/s: 669 rss: 95Mb L: 875/1567 MS: 2 ChangeBinInt-EraseBytes-
#2048	pulse  cov: 7151 ft: 29971 corp: 536/166Kb lim: 1577 exec/s: 682 rss: 95Mb
#2225	NEW    cov: 7151 ft: 29979 corp: 537/166Kb lim: 1577 exec/s: 741 rss: 95Mb L: 152/1567 MS: 2 ChangeBit-ChangeBit-
F
failures:

failures:
    communications::messages::message::tests::merge

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 353 filtered out

==30473== ERROR: libFuzzer: fuzz target exited
SUMMARY: libFuzzer: fuzz target exited
MS: 2 InsertByte-ChangeByte-; base unit: a5cab92cdd25d4d8fbe849e1c08cf3c9da97daec
artifact_prefix='/home/cfkaran2/Documents/repositories/bit_network/bit_network_library/src/communications/messages/__fuzz__/communications__messages__message__tests__merge/crashes/'; Test unit written to /home/cfkaran2/Documents/repositories/bit_network/bit_network_library/src/communications/messages/__fuzz__/communications__messages__message__tests__merge/crashes/crash-b90311f056ed895f4fd4ab9374f090e292023fa6
error: process exited with status 77
```

It's the exit code that I'm looking at; it's defined within LLVM itself in [FuzzerFlags.def](https://github.com/llvm/llvm-project/blob/master/compiler-rt/lib/fuzzer/FuzzerFlags.def).  So I suspect that either my code, or bolero, are somehow triggering this error.  I'm going to create a new bug on the LLVM project, and reference this bug.  Maybe we can all get to the bottom of this somehow!

**Comment by ckaran** [comment]
Bug is now filed at https://bugs.llvm.org/show_bug.cgi?id=46468

**Comment by ckaran** [comment]
I tried to create a reduced test, but now things are weirder than ever.  Copying a failing test and its corpora into a different project doesn't fail, but it continues to fail in the original project, even after I've fully cleaned everything and rebuilt from scratch.  This is an evil bug.

**Comment by camshaft** [comment]
Not really sure what's going on here. It not being reproducable would indicate some weird state the original project is in, but I'm not sure. Keep me posted on what you find!

**Comment by ckaran** [comment]
I suspect I've found something related to this bug, and I've created a reduced test for it.  See MR #23, in the `issues/reduced_bolero_test` directory.

That said, I don't think that fixing that bug will solve this bug entirely; it feels like that there is more going on here.

**Comment by ckaran** [comment]
Thank you for the new release (#47), it's already allowed me to catch some new errors that were being masked (fixes #27 & #28)!
